packages/lit-html/src/directives/guard.ts

```diff
@@ -0,0 +1,90 @@
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {
+  directive,
+  Directive,
+  noChange,
+  nothing,
+  Part
+} from '../lit-html.js';
+
+class Guard extends Directive {
+  previousValue: unknown = nothing;
+
+  render(_value: unknown, f: () => unknown) {
+    return f();
+  }
+
+  update(_part: Part, [value, f]: Parameters<this['render']>) {
```

Comment:
If we make the API variadic, we can simplify the implementation a little. We'd no longer need to check for array or non-array, it's just always an array of values.

Reply:
The API & name changes would probably be more of a breaking change than we're comfortable with since the public API is just following the decisions made in 1.x.

I do agree `memoize` is a more obvious naming of what the directive does.

Re: the parameters to memoize against, I recall discussing this before the 1.x release and believe there was pushback on using flat vargs due to the template function being a different type of arg, and an aversion to using vargs where e.g. "first/last arg is special".

Reply:
I also think passing the values as var args makes sense. It's pretty convenient with a rest arg as well. Perhaps that can be addressed in a separate PR that refines the directive API.

Reply:
I think we're talking about two difference things here. One is whether `Directive.update()` is variadic, and the other is whether `guard()` is.

I think @jridgewell is saying that if we made the `guard()` API _non-variadic_ - required the values to always be an array - that we could eliminate the `Array.isArray()` check. I agree on that. If I remember, we added arrays of values only later, so we preserved the single-value form for backwards compatibility. The only question for me is breaking change budget and the subtleness of the potential break when giving iterables as the single-value. If a string is passed as the value, we could easily iterate over each character on each render. We'd probably still want the `Array.isArray` check as an assertion.

For the `update()` API, I worry that making it variadic would lock us into these and only these arguments. For instance, we couldn't safely add an options argument. We'd have to pipe in any additional information via the part or another call like the constructor. Of course `render()` is variadic, with the idea that we want that call to have less information, and it makes it much easier for the author to automatically declare the signature of the directive function.

I can open separate issues for both things.